### ISSUE TITLE
Godror asks odpi-c to create a connection pool but does not use it

### ISSUE THREAD

**Issue by cvigo** [godror/godror#233]
**Describe the bug**
When the go sql library creates a new connection to a database, it goes through the funcion `Connect` here:
https://github.com/godror/godror/blob/9f1f724a508d1f088656c8cf5840cafd0a4f127f/drv.go#L1085
which, unless `connectionClass=NO-CONNECTION-POOLING` is set, gets a connection pool from the ODPI-C driver, here
https://github.com/godror/godror/blob/9f1f724a508d1f088656c8cf5840cafd0a4f127f/drv.go#L778

However, this pool is never used. No calls are ever made to `dpiPool_acquireConnection` as instructed in the [ODPI-C docs](https://oracle.github.io/odpi/doc/functions/dpiPool.html) and new connections outside the pool are created when needed. The code path always ends here:
https://github.com/godror/godror/blob/9f1f724a508d1f088656c8cf5840cafd0a4f127f/drv.go#L574

When the ODPI-C pool is "used", and a relevant number of new connections are needed in a very short period of time, they are created but not used and all the queries go through the first one that is created, so they run in sequence, not in parallel. Not sure if this weird effect is caused by the issue above.

**To Reproduce**
I created a test scenario in which hundreds of dummy queries are executed in parallel to a database which has no connections yet. As a result, connections are created, as many as concurrent queries, but then all queries go through one connection. In the following tcpdump you can see that one socket gets 419 packets while all others get 35-40

<img width="584" alt="image" src="https://user-images.githubusercontent.com/4451828/164272697-7130b04c-4ac1-45fc-86d7-3772b634425c.png">

The socket that runs all the queries
<img width="1115" alt="image" src="https://user-images.githubusercontent.com/4451828/164275723-b34c0bc6-f9de-482a-9330-b4f7d03edb60.png">

If I deactivate the ODPI-C pool, just passing `connectionClass=NO-CONNECTION-POOLING` to the connection string, all goes back to normal, and each one of the concurrent queries run on its own connection, all in parallel.

**Your oracle client version**
19.3

**Your godror version**
v0.33.0

**Your go version**
go1.17.linux-amd64

**Your gcc version**
e.g. 4.8.3

**Machine (please complete the following information):**
 - OS: Linux
 - Architecture: x86_64
 - Version: RHEL 8.4

**Comment by tgulacsi** [comment]
Yes, it always calls dpiCinn_create, but sets createParams.pool at https://github.com/godror/godror/blob/9f1f724a508d1f088656c8cf5840cafd0a4f127f/drv.go#L552

The doc says
"createParams [IN] – a pointer to a dpiConnCreateParams structure which is used to specify parameters for connection creation. NULL is also acceptable in which case all default parameters will be used when creating the connection. If the member pool is not NULL, a connection will be acquired from the pool (as if the function dpiPool_acquireConnection() had been called); otherwise, a standalone connection will be created."

So it should work, at least the problem is not here.

GT
________________________________
From: cvigo ***@***.***>
Sent: Wednesday, April 20, 2022 6:17:18 PM
To: godror/godror ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [godror/godror] Godror asks odpi-c to create a connection pool but does not use it (Issue #233)


Describe the bug
When the go sql library creates a new connection to a database, it goes through the funcion Connect here:
https://github.com/godror/godror/blob/9f1f724a508d1f088656c8cf5840cafd0a4f127f/drv.go#L1085
which, unless connectionClass=NO-CONNECTION-POOLING is set, gets a connection pool from the ODPI-C driver, here
https://github.com/godror/godror/blob/9f1f724a508d1f088656c8cf5840cafd0a4f127f/drv.go#L778

However, this pool is never used. No calls are ever made to dpiPool_acquireConnection as instructed in the ODPI-C docs<https://oracle.github.io/odpi/doc/functions/dpiPool.html> and new connections outside the pool are created when needed. The code path always ends here:
https://github.com/godror/godror/blob/9f1f724a508d1f088656c8cf5840cafd0a4f127f/drv.go#L574

When the ODPI-C pool is "used", and a relevant number of new connections are needed in a very short period of time, they are created but not used and all the queries go through the first one that is created, so they run in sequence, not in parallel. Not sure if this weird effect is caused by the issue above.

To Reproduce
I created a test scenario in which hundreds of dummy queries are executed in parallel to a database which has no connections yet. As a result, connections are created, as many as concurrent queries, but then all queries go through one connection. In the following tcpdump you can see that one socket gets 419 packets while all others get 35-40

[image]<https://user-images.githubusercontent.com/4451828/164272697-7130b04c-4ac1-45fc-86d7-3772b634425c.png>

The socket that runs all the queries
[image]<https://user-images.githubusercontent.com/4451828/164275723-b34c0bc6-f9de-482a-9330-b4f7d03edb60.png>

If I deactivate the ODPI-C pool, just passing connectionClass=NO-CONNECTION-POOLING to the connection string, all goes back to normal, and each one of the concurrent queries run on its own connection, all in parallel.

Your oracle client version
19.3

Your godror version
v0.33.0

Your go version
go1.17.linux-amd64

Your gcc version
e.g. 4.8.3

Machine (please complete the following information):

  *   OS: Linux
  *   Architecture: x86_64
  *   Version: RHEL 8.4

**Comment by sudarshan12s** [comment]
@cvigo, I think you are referring to the connections created as part of pool create in case of non-DRCP. 

The number of connections pre-created (poolMinSessions=20 ) in dpiPool_create will be reused for running the queries if you don't set connectionClass. 
example: 

`sql.Open("godror", ` user="demo" password="demo"   connectString="db1" poolMinSessions=20  poolMaxSessions=20 `)
`

It would cause all the 20 connections created as part of pool create to be used when Queries are executed.

**Comment by cvigo** [comment]
Hi @sudarshan12s Not sure if there is any practical difference between using the DRCP or a pool of permanent connections managed by Go `database/sql`. We use `(*DB).SetMaxIdleConns(x)` in any case, to make sure that a minimum of `x` connections are kept for reuse.

In my test case, I launch 200 queries in parallel, and 200 are executed through one connection, sequentially. I also get another 199 connections that established but do not perform any work. This seems abnormal to me.

**Comment by cvigo** [comment]
Thanks @tgulacsi, I certainly missed that point. So I don't know what else can be happening

**Comment by sudarshan12s** [comment]
> Hi @sudarshan12s Not sure if there is any practical difference between using the DRCP or a pool of permanent connections managed by Go `database/sql`. We use `(*DB).SetMaxIdleConns(x)` in any case, to make sure that a minimum of `x` connections are kept for reuse.
> 
> In my test case, I launch 200 queries in parallel, and 200 are executed through one connection, sequentially. I also get another 199 connections that are established but do not perform any work. This seems abnormal to me.

Can you share a minimal test case , It could be the connectionClass setting which may be causing issue. TestConcurrency in z_conc_test.go issues the queries in parallel and it displays pool stats like showing parallel connections. In tcpdump/wireshark, I locally can see multiple client ports sending insert queries, ..

> MaxOpenConnections:16 OpenConnections:16 InUse:16

**Comment by cvigo** [comment]
Hi @sudarshan12s 

Test tool here https://github.com/cvigo/dbconntest/tree/no_forks (use the branch "no_forks")

Example command with weird behavior:
```shell
dbconntest query_txn \
  --URL="user=XXXX password=yyyyy connectString=tcp://my-oracle-rac:1521/myDb" \
  --driver="godror" \
  --timeout=10s -\
  -sql="select to_char(current_timestamp) from dual" \
  --connections=200
```

But this works ok:
```shell
dbconntest query_txn 
  --URL="user=XXXX password=yyyyy connectString=tcp://my-oracle-rac:1521/myDb connectionClass=NO-CONNECTION-POOLING" \
  --driver="godror" \
  --timeout=10s \
  --sql="select to_char(current_timestamp) from dual" \
  --connections=200
```

Simplified test case here

```go
package main

import (
	"context"
	"database/sql"
	"sync"
	"time"
	_ "github.com/godror/godror"
)

func main() {

	db, err := sql.Open("godror", "user=XXXXXX password=YYYYYY connectString=tcp://my-oracle-db:1771/my-db-id")
	//BUT THIS WORKS => db, err = sql.Open("godror", "user=XXXXXX password=YYYYYY connectString=tcp://my-oracle-db:1771/my-db-id connectionClass=NO-CONNECTION-POOLING")
	if err != nil {
		panic("Error connecting to DB")
	}

	db.SetMaxIdleConns(0) // Not interested in connection reuse for this type of test. Happy to drop after the operation ends
	db.SetMaxOpenConns(300)

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// spread the workers
	waitGroup := &sync.WaitGroup{}
	waitGroup.Add(200)
	for i := 0; i < 200; i++ {
		go func(i int) {
			doWork(ctx,  db)
			waitGroup.Done()
		}(i)
	}
	waitGroup.Wait()
	_ = db.Close()
}

func doWork(ctx context.Context, db *sql.DB) error {
	tx, err := db.BeginTx(ctx, &sql.TxOptions{
		Isolation: sql.LevelDefault,
		ReadOnly:  false,
	})
	if err != nil {
		return err
	}
	rows, err := tx.QueryContext(ctx, "select to_char(current_timestamp) from dual")
	if err != nil {
		return err
	}
	err = tx.Commit()
	_ = rows.Close()
	return err
}
```

**Comment by sudarshan12s** [comment]
> Hi @sudarshan12s
> 
> Test tool here https://github.com/cvigo/dbconntest/tree/no_forks (use the branch "no_forks")
> 
> Example command with weird behavior:
> 
> ```shell
> dbconntest query_txn \
>   --URL="user=XXXX password=yyyyy connectString=tcp://my-oracle-rac:1521/myDb" \
>   --driver="godror" \
>   --timeout=10s \
>   --sql="select to_char(current_timestamp) from dual" \
>   --connections=200
> ```
> 
> But this works ok:
> 
> ```shell
> dbconntest query_txn \
>   --URL="user=XXXX password=yyyyy connectString=tcp://my-oracle-rac:1521/myDb connectionClass=NO-CONNECTION-POOLING" \
>   --driver="godror" \
>   --timeout=10s \
>   --sql="select to_char(current_timestamp) from dual" \
>   --connections=200
> ```
> 
> Simplified test case here
> 
> ```go
> package main
> 
> import (
> 	"context"
> 	"database/sql"
> 	"sync"
> 	"time"
> 	_ "github.com/godror/godror"
> )
> 
> func main() {
> 
> 	db, err := sql.Open("godror", "user=XXXXXX password=YYYYYY connectString=tcp://my-oracle-db:1771/my-db-id")
> 	//BUT THIS WORKS => db, err = sql.Open("godror", "user=XXXXXX password=YYYYYY connectString=tcp://my-oracle-db:1771/my-db-id connectionClass=NO-CONNECTION-POOLING")
> 	if err != nil {
> 		panic("Error connecting to DB")
> 	}
> 
> 	db.SetMaxIdleConns(0) // Not interested in connection reuse for this type of test. Happy to drop after the operation ends
> 	db.SetMaxOpenConns(300)
> 
> 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
> 	defer cancel()
> 
> 	// spread the workers
> 	waitGroup := &sync.WaitGroup{}
> 	waitGroup.Add(200)
> 	for i := 0; i < 200; i++ {
> 		go func(i int) {
> 			doWork(ctx,  db)
> 			waitGroup.Done()
> 		}(i)
> 	}
> 	waitGroup.Wait()
> 	_ = db.Close()
> }
> 
> func doWork(ctx context.Context, db *sql.DB) error {
> 	tx, err := db.BeginTx(ctx, &sql.TxOptions{
> 		Isolation: sql.LevelDefault,
> 		ReadOnly:  false,
> 	})
> 	if err != nil {
> 		return err
> 	}
> 	rows, err := tx.QueryContext(ctx, "select to_char(current_timestamp) from dual")
> 	if err != nil {
> 		return err
> 	}
> 	err = tx.Commit()
> 	_ = rows.Close()
> 	return err
> }
> ```

Thanks . Can you try setting poolMinSessions to some reasonable value to get initialised in pool and see if it helps. 

> db, err := sql.Open("godror", "user=XXXXXX password=YYYYYY connectString=tcp://my-oracle-db:1771/my-db-id poolMinSessions=20" )

**Comment by cvigo** [comment]
hi @sudarshan12s just tried `poolMinSession=200`. Some race happens as I saw 600 tcp connections to Oracle and I got 200 `ORA-02391: exceeded simultaneous SESSIONS_PER_USER limit` errors :-(

**Comment by sudarshan12s** [comment]
> hi @sudarshan12s just tried `poolMinSession=200`. Some race happens as I saw 600 tcp connections to Oracle and I got 200 `ORA-02391: exceeded simultaneous SESSIONS_PER_USER limit` errors :-(
> 
> Leaving the pool management to the DRCP does not seem to be a good idea...

when you set poolMinSession=200, the local session pool gets created/initialized with 200 outbound connections and hence 200 sessions which could be crossing the session limits on server, hence failing pool creation. 

Generally these outbound connections are shared from multiple go routines. Normally  a goroutine would not hold connection long time but returns to pool after the DB query which can be used from another goroutine. So at any instant , number of goroutines concurrently holding the DB connections would be generally less. If more such concurrent requests are coming, OCI session pool expands at runtime till poolMaxSessions . poolIncrement is used during expansion.  so poolMinSession is just starting number  of outbound connections when pool is created. 

These settings are for the local session pool created at OCI library at client side. Its not with the server side DRCP. DRCP is not used in your case as i can see from your example.

**Comment by cvigo** [comment]
Just to clarify why I am doing this and why I would not like to prepare a pool of connections beforehand.

We are setting up a Kubernetes Cluster which will run thousands of Services as containers. The services connect to databases that are managed by a single Oracle RAC. We cannot allow the Services to connect directly to Oracle as the number of connections would be huge (and each connections spawns a process in Oracle, so we can even exceed the maximum 10K process our RHEL servers can handle).

Therefore we have developed a middleware that gets the SQL commands through gRPC from the services (which is also useful to get rid off the Oracle drivers in the Kubernetes Services) and sends them to the Oracle Server. This middleware handles the connections to Oracle in a lazy manner: just delegates the pool management to `database/sql library`. Depending on the amount of traffic each database gets, the number of connections in the pool increases.

The middleware has no means to calculate how much traffic each database will get, so it can hardly preset a good minimum number of connections before accepting incoming SQL request from the Services.

Some databases get 2000-3000 sql per second so, when one of the nodes of the middleware is restarted due to failure or maintenance, it gets gets hundreds of requests in a very short period of time.

**Comment by cvigo** [comment]
> when you set poolMinSession=200, the local session pool gets created/initialized with 200 outbound connections and hence 200 sessions which could be crossing the session limits on server, hence failing pool creation.

In the latest test I had 200 goroutines trying to create a pool concurrently (I have observed 200 calls to `dpiPool_create` calls). Apparently each call tries to create 200 for sure I overrun the maximum allowed

**Comment by tgulacsi** [comment]
You have Thundering Herd problem, with a C library that does concurrency badly.
So, lower your poolMinSessions to 20 for example (as @sudarshan12s suggested).

Another mitigation is to limit the concurrency for startup: either by putting a golang.org/x/time/rate.Limiter before db actions,
or by allowing only ONE goroutine to initialize the connection pool - or do it ASAP at application startup with a db.Ping, BEFORE starting the gRPC server. This has an additional benefit: it will test&bail out on bad config early, before the clients are allowed to connect.

**Comment by sudarshan12s** [comment]
> > when you set poolMinSession=200, the local session pool gets created/initialized with 200 outbound connections and hence 200 sessions which could be crossing the session limits on server, hence failing pool creation.
> 
> In the latest test I had 200 goroutines trying to create a pool concurrently (I have observed 200 calls to `dpiPool_create`). Apparently each call tries to create 200 for sure I overrun the maximum allowed

There could be a race, hence db.Ping() in main goroutine needs to be added to get pool created before spawning the goroutines. Hence subsequent goroutines spawned just acquire connections from already created pool.

**Comment by cvigo** [comment]
> or by allowing only ONE goroutine to initialize the connection pool - or do it ASAP at application startup with a db.Ping, BEFORE starting the gRPC server. This has an additional benefit: it will test&bail out on bad config early, before the clients are allowed to connect.

> There could be a race, hence db.Ping() in main goroutine needs to be added to get pool created before spawning the goroutines. Hence subsequent goroutines spawned just acquire connections from already created pool.

Agree and actually that's the classic solution. The point here is: how big do I make the pool. Some databases managed by the middleware get 3000 SQL/s and others get 3000 SQL/month. I will have to do some guessing and adjust from time to time :-(

**Comment by sudarshan12s** [comment]
poolIncrement will control on number of more outbound connections to be used to expand the pool dynamically as part of single request and all existing connections are busy.  

poolSessionTimeout , poolSessionMaxLifeTime would cause dynamic pool shrinking till poolMinSession value, there by freeing the resources. 

I guess DRCP along with client session pool might help in your use case .

> sql.Open("godror", ` user="demo" password="demo" connectionClass=TestMyClass  connectString="ip:1521/db1:POOLED" poolSessionTimeout=300s poolWaitTimeout=3000000 `)

**Comment by sudarshan12s** [comment]
some info on DRCP: 

With DRCP, the clients are connected to the broker process on the database host, and share a fewer number of pooled server processes for doing the SQL activity. The connections to the broker can be much larger than the number of server processes. The server process gets allocated when the client wants a query to be run and requests for a server.  Please see these links
 
https://www.oracle.com/pls/topic/lookup?ctx=dblatest&id=GUID-BB76E57C-3F16-4C85-AEF6-BA14FC1B4777
 
https://www.oracle.com/technetwork/topics/php/php-scalability-ha-twp-128842.pdf

**Comment by cvigo** [comment]
Thanks for that @sudarshan12s 

This will help to mitigate another issue we have (100 classic JBoss apps, replicated in 100 clones,  accessing Oracle through JDBC and therefore hitting the 10000 max process a Oracle Linux server can handle.

For my issue in my middleware, I think I have no better choice that allocating the pools in my middleware before opening the service my consumers. This will require the calling apps to reuse the same (or a few) user IDs, so I have to deal with a small number of pools.
